vault/identity_store_entities.go

```diff
@@ -0,0 +1,463 @@
+package vault
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/fatih/structs"
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/vault/helper/locksutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"github.com/mitchellh/mapstructure"
+)
+
+// entityPaths returns the API endpoints supported to operate on entities.
+// Following are the paths supported:
+// entity - To register a new entity
+// entity/id - To lookup, modify, delete and list entities based on ID
+// entity/merge/id - To merge entities based on ID
+func entityPaths(i *identityStore) []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: "entity",
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the entity",
+				},
+				"metadata": {
+					Type:        framework.TypeMap,
+					Description: "Metadata to be associated with the entity",
+				},
+				"policies": {
+					Type:        framework.TypeStringSlice,
+					Description: "Policies to be tied to the entity",
+				},
+				// "identities", extracted from raw field data
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: i.pathEntityRegister,
+			},
+
+			HelpSynopsis:    strings.TrimSpace(entityHelp["entity"][0]),
+			HelpDescription: strings.TrimSpace(entityHelp["entity"][1]),
+		},
+		{
+			Pattern: "entity/id/" + framework.GenericNameRegex("id"),
+			Fields: map[string]*framework.FieldSchema{
+				"id": {
+					Type:        framework.TypeString,
+					Description: "ID of the entity",
+				},
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the entity",
+				},
+				"metadata": {
+					Type:        framework.TypeMap,
+					Description: "Metadata to be associated with the entity",
+				},
+				"policies": {
+					Type:        framework.TypeStringSlice,
+					Description: "Policies to be tied to the entity",
+				},
+				// "identities", extracted from raw field data
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: i.pathEntityIDUpdate,
+				logical.ReadOperation:   i.pathEntityIDRead,
+				logical.DeleteOperation: i.pathEntityIDDelete,
+			},
+
+			HelpSynopsis:    strings.TrimSpace(entityHelp["entity-id"][0]),
+			HelpDescription: strings.TrimSpace(entityHelp["entity-id"][1]),
+		},
+		{
+			Pattern: "entity/id/?$",
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ListOperation: i.pathEntityIDList,
+			},
+
+			HelpSynopsis:    strings.TrimSpace(entityHelp["entity-id-list"][0]),
+			HelpDescription: strings.TrimSpace(entityHelp["entity-id-list"][1]),
+		},
+		{
+			Pattern: "entity/merge/id/?$",
```

Comment:
The idea behind having `id` in the path was to indicate that the merge is happening based on IDs. It can very well happen on other factors if needed, going forward. It is intended to be only a placeholder. But I see your point. It is odd to have `id` in path and a value not expected afterwards.
OR
I could introduce a field specifier indicating the criteria for merge. This seems like an over kill too.

_is inclined towards having `id` in path_

Let me know your thoughts.